### PR TITLE
Patch/upgrade bsp

### PR DESCRIPTION
**Pull request by noslav**
In order to run only against the most recent blocks aka live sync historical sync has to be avoided (Historical mode: 0, Live mode: 1..). Keeping this as a `*uint32` to expand use cases in the future to other modes 

1. Set up a thread to monitor at start of node, to check if the historical synchronization is complete
2. Atomically set the block replica config flag var for when the historical sync is finished using the handler interface `Synced()` fn
3. Read the atomically set value to determine if the most recent blocks should be exported accordingly in `block_replica.go`
4. Set the default values of replica config historical flag to 0 when starting a node so only the txs for recent blocks are sent (without overwhelming the proof-chain with txs that have been missed for earlier blocks)

### CODE REVIEW

File: les/api_backend.go [hunk 1]
```diff
@@ -331,3 +332,13 @@ func (b *LesApiBackend) StateAtBlock(ctx context.Context, block *types.Block, re
 func (b *LesApiBackend) StateAtTransaction(ctx context.Context, block *types.Block, txIndex int, reexec uint64) (core.Message, vm.BlockContext, *state.StateDB, error) {
 	return b.eth.stateAtTransaction(ctx, block, txIndex, reexec)
 }
+
+func (b *LesApiBackend) SetHistoricalBlocksSynced() bool {
+	if b.eth.handler.downloader.Progress().CurrentBlock < b.eth.handler.downloader.Progress().HighestBlock {
```
Comment (sudeepdino008) [inline les/api_backend.go:337]: can use b.SyncProgress() == HighestBlock etc.

File: les/api_backend.go [hunk 2]
```diff
@@ -331,3 +332,13 @@ func (b *LesApiBackend) StateAtBlock(ctx context.Context, block *types.Block, re
...
+		log.Info("Light Client not fully synced yet and block specimen producer does not operate in light client mode")
```
Comment (sudeepdino008) [inline les/api_backend.go:338]: should be error log?

File: les/api_backend.go [hunk 3]
```diff
@@ -331,3 +332,13 @@ func (b *LesApiBackend) StateAtBlock(ctx context.Context, block *types.Block, re
...
+		return false
+	} else {
+		log.Info("Fully Synced but block specimen producer does not operate in light client mode", "Current block", b.eth.handler.downloader.Progress().CurrentBlock, "Highest block", b.eth.handler.downloader.Progress().HighestBlock)
```
Comment (sudeepdino008) [inline les/api_backend.go:341]: in fact we should probably fatal out here given light client mode is not supported.

File: eth/api_backend.go [hunk 4]
```diff
@@ -359,3 +361,15 @@ func (b *EthAPIBackend) StateAtBlock(ctx context.Context, block *types.Block, re
...
 }
+
+// SetHistoricalBlocksSynced returns a bool for BSP replica config (Historical mode :0 , Live mode: 1)
+func (b *EthAPIBackend) SetHistoricalBlocksSynced() bool {
+	if b.eth.Synced() {
+		atomic.StoreUint32(b.eth.ReplicaConfig.HistoricalBlocksSynced, 1)
+		log.Info("Fully Synced, BSP running in live sync mode", "BSP Mode Config: ", atomic.LoadUint32(b.eth.ReplicaConfig.HistoricalBlocksSynced))
+		return true
+	} else {
+		log.Info("Not accepting new transactions, BSP running in historical sync mode", "BSP Mode Config: ", atomic.LoadUint32(b.eth.ReplicaConfig.HistoricalBlocksSynced))
```
[the remark this answers is not in the document]
Comment (noslav) [reply eth/api_backend.go:372]: agreed changing to `log.Error`

File: les/api_backend.go (hunk 3 above)
Comment (noslav) [reply les/api_backend.go:341]: yes agreed this should fatal out

Comment (noslav) [reply les/api_backend.go:341]: moving to `log.Crit`

File: cmd/geth/main.go [hunk 5]
```diff
@@ -433,6 +433,31 @@ func startNode(ctx *cli.Context, stack *node.Node, backend ethapi.Backend, isCon
 			utils.Fatalf("Failed to start mining: %v", err)
 		}
 	}
+
+	// Spawn a standalone goroutine for status synchronization monitoring,
+	// if full sync is completed in block specimen creation mode set replica config flag
+	if ctx.GlobalBool(utils.ReplicaEnableSpecimenFlag.Name) || ctx.GlobalBool(utils.ReplicaEnableResultFlag.Name) {
+		log.Info("Synchronisation started, historical blocks synced set to 0")
+		go func() {
+			sub := stack.EventMux().Subscribe(downloader.DoneEvent{})
```
Comment (sudeepdino008) [inline cmd/geth/main.go:442]: according to the struct TypeMux documentation, it is deprecated and should be replaced by Feeds.
See:
https://github.com/ethereum/go-ethereum/pull/16706
https://github.com/ethereum/go-ethereum/pull/20312/

Most places in geth still use TypeMux though and it works, so I would leave this up to you if you want to implement this.

Comment (noslav) [reply cmd/geth/main.go:442]: why not take that up as a ticket?

**Review by sudeepdino008 — APPROVED** [review]

**Pull request merged**